File: 1268_suggestedProducts.py
from heapq import heappush, nsmallest

class TrieNode:
    def __init__(self):
        self.children = {}
        self.index = []
    
class Trie:
    def __init__(self):
        self.root = TrieNode()
        
    def get_root(self):
        return self.root
    
    def insert(self, word, index):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            heappush(node.children[char].index, index)
            node = node.children[char]
        
    def find(self, node, char):
        if char in node.children:
            result = nsmallest(3, node.children[char].index)
            return node.children[char], result
        else:
            return None, []   
    
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        trie = Trie()
        for i, word in enumerate(products):
            trie.insert(word, i)
        root = trie.get_root()
        output = []
        for i, char in enumerate(searchWord):
            root, result = trie.find(root, char)
            if result:
                temp = []
                for idx in result:
                    temp.append(products[idx])
                output.append(temp)
            else:
                output.append([])
                break
        output += [[]] * (len(searchWord) - i - 1)
        return output

# Alternative solution

class TrieNode:
    def __init__(self):
        self.children = {}
        self.suggestions = []
        
class Trie:
    def __init__(self):
        self.root = TrieNode()

    def get_root(self):
        return self.root

    def insert(self, word):
        node = self.root

        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()

            node = node.children[char]

            if len(node.suggestions) < 3:
                node.suggestions.append(word)

    def suggest(self, node, char):
        if node and char in node.children:
            return node.children[char], node.children[char].suggestions 
        return None, []

class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        trie = Trie()

        for product in products:
            trie.insert(product)

        output = []
        root = trie.get_root()

        for char in searchWord:
            if not root:
                output.append([])
            else:
                root, suggestion = trie.suggest(root, char)
                output.append(suggestion)

        return output

File: 1001-1500/1268_suggestedProducts.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        int n = products.length;
        List<List<String>> output = new ArrayList<>();
        int start = 0;
        int index = 1;

        while (start < n && index <= searchWord.length()) {
            String s = products[start];
            if (!s.substring(0, Math.min(index, s.length())).equals(searchWord.substring(0, index))) {
                start++;
            } else {
                List<String> suggest = new ArrayList<>();
                suggest.add(s);

                if (start + 1 < n) {
                    String s1 = products[start + 1];
                    if (s1.substring(0, Math.min(index, s1.length())).equals(searchWord.substring(0, index))) {
                        suggest.add(s1);
                    }
                }
                if (start + 2 < n) {
                    String s2 = products[start + 2];
                    if (s2.substring(0, Math.min(index, s2.length())).equals(searchWord.substring(0, index))) {
                        suggest.add(s2);
                    }
                }
                output.add(suggest);
                index++;
            }
        }

        int l = output.size();

        for (int i = 0; i < searchWord.length() - l; i++) {
            output.add(new ArrayList<>());
        }

        return output;
    }
}

// Alternative solution

class TrieNode {
    HashMap<Character, TrieNode> children;
    List<String> suggestions;
    boolean isEnd;

    public TrieNode() {
        this.children = new HashMap<>();
        this.suggestions = new ArrayList<>();
        this.isEnd = false;
    }
}


class Trie {
    TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }
    
    public void insert(String word) {
        TrieNode node = this.root;

        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (!node.children.containsKey(c)) {
                node.children.put(c, new TrieNode());
            }
            node = node.children.get(c);
            if (node.suggestions.size() < 3) {
                node.suggestions.add(word);
            }
        }

        node.isEnd = true;
    }
    
}

class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        Trie trie = new Trie();
        
        for (String product: products) {
            trie.insert(product);
        }

        List<List<String>> suggestions = new ArrayList<>();
        TrieNode node = trie.root;

        for (int i = 0; i < searchWord.length(); i++) {
            char c = searchWord.charAt(i);
            if (node != null && node.children.containsKey(c)) {
                node = node.children.get(c);
                suggestions.add(node.suggestions);
            } else {
                node = null;
                suggestions.add(new ArrayList<>());
            }
        }

        return suggestions;
    }
}
